src/inmanta/ast/statements/generator.py

```diff
@@ -79,13 +83,25 @@ def __init__(self, instance_type: "Entity", implements: "Implement") -> None:
         self.implements = implements
 
     def normalize(self) -> None:
-        # done in define type
-        pass
+        injected_variables: Set[str] = {"self"}.union(self.type.get_all_attribute_names())
+        self._own_eager_promises = [
+            # implementations live in the namespace's context rather than the constructor's context so for promises that cross
+            # the boundary we translate references so that they are resolved correctly in any context wrapping the constructor.
+            dataclasses.replace(promise, instance=promise.instance.fully_qualified())
+            for implementation in self.implements.implementations
+            for promise in implementation.statements.get_eager_promises()
+            if promise.get_root_variable() not in injected_variables
+        ]
+        # TODO: temporary override because implementation blocks have not normalized yet at this point, meaning the above is
```

Comment:
this I don't get.

Reply:
This is about the normalization dependency loop we discussed yesterday. The list comprehension above is accurate if everything below it has already normalized but with the current normalization order it hasn't. Still need to look further into the options we talked about yesterday.